README.md

```diff
@@ -1,10 +1,260 @@
+   the Dockerfile or its dependencies change.
+* `make tests` - launch all the tests. Worth noting is the fact that mccabe
+  complexity of the code is also verified and an error is raised if it's equal
+  or above 10.
+* `make shell` - launch an interactive shell within the devkit container. Should
+  be used in case when fine grained control of the tests is necessary or
+  during debugging.
+* `make flake8` - launch flake8 which will check all the tests and test-harness
+  files by default.
+
+### Docker container
+As mentioned earlier, all the commands are executed inside the `adminrouter-devkit`
+container. It follows the same build process for Nginx that happens during
+DC/OS build with the execption of setting the  `--with-debug` flag. It also
+contains some basic debugging tools, pytest related dependencies and basic
+files that help pytest mimic the DC/OS environment. Their location is then
+exported via environment variables to the pytest code, so changing their location
+can be done by edition only the Dockerfile.
+
+Due to some issues with overlayfs on older kernels, the container links
+/tmp directory to /dev/shm.
+
+### Repository flavours
+Admin Router repository can come in two variants/flavours - Enterprise and Opensource.
+Tests determine it basing on the directory structure - if tests contain
+`test-harness/tests/open/` directory then the repo is treated as Opensource one,
+and in case when `test-harness/tests/ee/` directory is present - as the
+enterprise one. Pytest fixture `repo_is_ee` takes care of it and is pulled in
+as dependencies by all the fixtures that need this information.
+
+### Service startup ordering and cleanup
+Mocking out DC/OS is a complex task, and a huge part in it have pytest
+fixtures. They take care of proper ordering of mocks and subprocess start/stop and
+cleanup and the cleanup ordering itself.
+
+Tracking the chain of fixtures and how they use each other may provide better
+understanding of how the test harness works.
+
+### JWT
+DC/OS IAM relies on JSON Web Tokens for transmitting security information between DC/OS
+components. Open repository flavour uses HS256 tokens, while EE relies on RS256.
+Test harness needs to have a way to create them for use in the tests. This is
+done by `test-harness/modules/mocker/jwt.py` module, which together with
+`repo_is_ee` fixture provides abstracts away the type of the
+token used by Admin Router itself.
+
+### Mocker
+Mocker takes care of simulating DC/OS HTTP endpoints that Admin Router uses. It's
+basically just a thin management layer on top of multiple python-based
+HTTP servers, each one of them mocking out particular component/upstream of DC/OS.
+
+It exposes the `.send_command()` method that allows to reconfigure endpoints
+the way the tests need it. It basically calls a getattr on the endpoint
+instance in order to execute given function and pass it attributes as specified
```

Comment:
It implements dynamic dispatch by calling the function of the given name on the endpoint instance, passing along the parameters.